ethcore/src/client/client.rs

```diff
@@ -1318,6 +1318,33 @@ impl BlockChainClient for Client {
 		self.do_virtual_call(&env_info, &mut state, &t, analytics)
 	}
 
+*/
+	fn replay(&self, id: TransactionId, analytics: CallAnalytics) -> Result<Executed, CallError> {
+		let address = self.transaction_address(id).ok_or(CallError::TransactionNotFound)?;
+		let block = BlockId::Hash(address.block_hash);
+
+		Ok(self.replay_block_transactions(block, analytics)?.nth(address.index).unwrap())
+	}
+
+	fn replay_block_transactions(&self, block: BlockId, analytics: CallAnalytics) -> Result<Box<Iterator<Item = Executed>>, CallError> {
+		let mut replays = Vec::new();
+		let mut env_info = self.env_info(block).ok_or(CallError::StatePruned)?;
+		let body = self.block_body(block).ok_or(CallError::StatePruned)?;
+		let mut state = self.state_at_beginning(block).ok_or(CallError::StatePruned)?;
+		let txs = body.transactions();
+
+		const PROOF: &'static str = "Transactions fetched from blockchain; blockchain transactions are valid; qed";
+		const EXECUTE_PROOF: &'static str = "Transaction replayed; qed";
+		for t in txs {
+			let t = SignedTransaction::new(t).expect(PROOF);
+			let x = self.do_virtual_call(&env_info, &mut state, &t, analytics).expect(EXECUTE_PROOF);
+			env_info.gas_used = env_info.gas_used + x.gas_used;
+			replays.push(x);
+		}
+
+		Ok(Box::new(replays.into_iter()))
```

Comment:
It's almost fine - right now you are executing all transactions and putting the results into a vector, but `Iterator` allows you to do it lazily (which will make `replay` work faster).
It should look like this (instead of a loop you should use `map`):
```
let txs = block.transactions();
Ok(Box::new(txs.into_iter()
  .map(|tx| {
     ...
     self.do_virtual_call(...).expect(EXECUTE_PROOF)
   })))
```